### PR TITLE
feat: detect traditional and integration tests

### PR DESCRIPTION
**Pull request by didrocks**
Add `MustBeTesting` assertion to ensure we run in tests. This assertion will panic if we don’t run in traditional tests (gotest)
or integration tests.

Add tests for detection tests
Those tests address:
- traditional tests detection started by go test
- integration tests detection for a binary built with the
  integrationtests tag.
- ensure we don’t falsy report integration tests for binary without the
  tag.

Hook up coverage report, even for the subprocess, with this.

UDENG-2563
UDENG-2584

### CODE REVIEW

**Review by denisonbarbosa — APPROVED** [review]
Nice work!

File: internal/testsdetection/testsdetection_test.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	t.Parallel()
+
+	tests := map[string]struct {
+		integrationtestsTag bool
+
+		wantPanic bool
+	}{
+		"Pass when called in an integration tests build": {integrationtestsTag: true, wantPanic: false},
+
+		"Error (panics) when called in non tests and no integration tests": {integrationtestsTag: false, wantPanic: true},
```
Comment (3v1n0) [inline internal/testsdetection/testsdetection_test.go:37]: Just a nit, but personally I find inline test case structs very hard to follow :stuck_out_tongue_closed_eyes:

File: internal/testsdetection/testsdetection_test.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+			testBinary := filepath.Join(temp, "testbin")
+
+			buildCmd := []string{"build", "-o", testBinary}
+			if tc.integrationtestsTag {
+				buildCmd = append(buildCmd, "-tags=integrationtests")
+			}
+			env := os.Environ()
+			if coverDir != "" {
+				buildCmd = append(buildCmd, "-cover")
+				env = append(env, "GOCOVERDIR="+coverDir)
```
Comment (3v1n0) [inline internal/testsdetection/testsdetection_test.go:53]: Nit, but we've a testutils definition to handle this no? (`AppendCovEnv` or `CoverDirEnv`)

Or do we want to keep tests here independent?

File: internal/testsdetection/testsdetection_test.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+			}
+			buildCmd = append(buildCmd, "testdata/binary.go")
+
+			//nolint:gosec // G204 we are in control of the arguments in our tests.
+			out, err := exec.Command("go", buildCmd...).CombinedOutput()
+			require.NoErrorf(t, err, "Setup: Could not build test binary: %s", out)
```
Comment (3v1n0) [inline internal/testsdetection/testsdetection_test.go:59]: Denison once told me he preferred the non `f` definitions for those

File: internal/testsdetection/testsdetection_test.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+			// Execute our subprocess
+			//nolint:gosec // G204 we are in control of the arguments in our tests.
+			cmd := exec.Command(testBinary)
+			cmd.Env = env
+			out, err = cmd.CombinedOutput()
```
Comment (3v1n0) [inline internal/testsdetection/testsdetection_test.go:65]: Why not just use `go run` instead of `go build` + execution?

File: internal/testsdetection/testsdetection_test.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+			if tc.wantPanic {
+				require.Errorf(t, err, "MustBeTesting should have panicked the subprocess: %s", string(out))
```
Comment (3v1n0) [inline internal/testsdetection/testsdetection_test.go:68]: no need to use `string(out)` if using `%s` already

**Review by 3v1n0 — COMMENTED** [review]
LGTM, see if want to consider few comments though :)

Also maybe use some prefix on commit titles? To make clearer what they're addressing.

File: internal/testsdetection/testsdetection_test.go (hunk 5 above)
Comment (didrocks) [reply internal/testsdetection/testsdetection_test.go:68]: That was my first attempt, but I got a slice of byte printed out here.

Comment (3v1n0) [reply internal/testsdetection/testsdetection_test.go:68]: Weird, I've been using that too recently and it did work.

File: internal/testsdetection/testsdetection_test.go (hunk 4 above)
Comment (didrocks) [reply internal/testsdetection/testsdetection_test.go:65]: `go run ` with `-cover` didn’t work in `1.20`, I just did an attempt and indeed, it works with 1.22, I will replace with this.

File: internal/testsdetection/testsdetection_test.go (hunk 3 above)
Comment (didrocks) [reply internal/testsdetection/testsdetection_test.go:59]: here, we are using %s, so we do have formatting. It makes sense to have the `f` here.

File: internal/testsdetection/testsdetection_test.go (hunk 2 above)
Comment (didrocks) [reply internal/testsdetection/testsdetection_test.go:53]: oh, good ones, I didn’t know we already had helpers for those, I will use them in a followup PR

File: internal/testsdetection/testsdetection_test.go (hunk 1 above)
Comment (didrocks) [reply internal/testsdetection/testsdetection_test.go:37]: On the contrary, I find it way easier and it describes a story :)

File: internal/testsdetection/testsdetection_test.go (hunk 4 above)
Comment (3v1n0) [reply internal/testsdetection/testsdetection_test.go:65]: It seems it does (at least if order is correct), as we use it in `TestPamCLIRunStandalone` and that generates the coverage: https://app.codecov.io/github/ubuntu/authd/blob/main/pam%2Fmain-cli.go#L146

File: internal/testsdetection/testsdetection_test.go (hunk 3 above)
Comment (3v1n0) [reply internal/testsdetection/testsdetection_test.go:59]: Yeah, I kinda agree, just reporting what Denison said to me, so I wonder if we have a team-preference for those cases :)

File: internal/testsdetection/testsdetection_test.go (hunk 5 above)
Comment (didrocks) [reply internal/testsdetection/testsdetection_test.go:68]: Ahah, found it! The issue is on https://github.com/ubuntu/authd/pull/300/files/531a090e7e64b14d6c4b188b9fa4b508aecedf57..29a6185a9e4c36a2a190aa63bcbf63d2d514619b#diff-c00785a26f57a11b8084dd9fc39d8bab7944c6db4c221fabd2abbd7578461561R71 which is missing `%s` and that was the one I saw when erroring out :)
